garage/tf/policies/base2.py

```diff
@@ -88,6 +94,25 @@ def terminate(self):
         """Clean up operation."""
         pass
 
+    def get_trainable_vars(self, scope=None):
+        """Get trainable vars."""
+        scope = scope if scope else self.name
```

Comment:
`self.name` is rarely the full scope, which is actually a path, right?

Reply:
e.g. my scope might be `Foo/Bar/Baz/Buz/str(self.name)`, but what you are doing is just `str(self.name)`

Reply:
Do you mean if the policy is wrapped by something else, it won't work as the scope will be different?

Reply:
we have no idea which outer scope the policy was created in. there's no reason i can't do

```
with LocalRunner():
    with tf.variable_scope('arbitrary'):
        policy = GaussianMLPPolicy(name='foo')  # scope is `arbitrary/foo`

    PPO(policy=policy, env=env)
```

Reply:
right. Perhaps getting variables under `tf.get_variable_scope` will do the job?

Reply:
uhh i'm not sure exactly what you mean, but if you are suggesting we search the variable_scope space then definitely not.

your method as-defined is a strange mix of an instance method and a class method, and i think it's really confusing.

for an instance method, you want to satisfy the signature:
```python
class Policy:

    """
    note there are no arguments (neither optional nor required) because 
    Policy should already know exactly which scope(s) its variables live in. if 
    MyPolicy doesn't know where its variables live, how on earth would i know???
    why would I define a Policy object at all if it's not going to keep track of that?
    """
    def get_trainable_vars(self):

        # magic

p = MyPolicy()
trainable = p.get_trainable_vars() 
```

```python
class Policy:

    """
    here we ask the user to keep track of the scope. this makes sense as a classmethod,
    though i'm not sure why anyone would actually want to use it this way (perhaps as a
    private helper for child classes)

    note scope is *required* because Policy either knows the scope or it doesn't. having
    something like "try to guess the scope if i don't tell you (and barf if you're wrong) but also
    use whatever scope i tell you even if it's wrong" is very surprising behavior.
    """
    @classmethod
    def get_trainable_vars(cls, scope):
        # magic
```

Reply:
tl;dr: your policy either needs to keep around a VariableScope object somehow or you need to figure out a different way of doing this

for instance, you can have `Model.get_trainable_vars()` and then have a policy concatenate all the variables of its models:

```python
class MyPolicy:

    def get_trainable_vars(self):
        vars = []
        vars.extend(self._model1.get_trainable_vars())
        vars.extend(self._model2.get_trainable_vars())

        return vars
```

If you wanted this to be magically-implemented by `Policy` for all policies, you would have to force the child classes to maintain some registry of models so that the parent class can query all of them
```python
class Policy:

    def get_trainable_vars(self):
        vars = []
        for m in self._models:
            vars.extend(m.get_trainable_vars())

        return vars

class MyPolicy(Policy):

    def __init__(self):
        model = MyModel()
        self._add_model(model)
```

Alternatively, you can make sure that there's an enclosing scope for each `Policy` in which all Models are built (this makes a lot of sense). You store that VariableScope, and then use it to query for anything your models created:

```python
class Policy:

    def __init__(self, name=None):
        self._variable_scope = tf.variable_scope(name)

    def _initialize(self):
        with self._variable_scope:
            # stuff

    def get_trainable_vars(self):
        return tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, scope=self._variable_scope)
```

this is probably the best option, since it's what will happen once we inherit `Policy` from `Model`

Reply:
It makes a lot of sense. The VariableScope of the `Model` will later become the VariableScope of the `Policy`.

Reply:
which option did you choose?

Reply:
Second approach, which is to build all models with the same VariableScope and used that to query for anything.

Reply:
do you mean that two models will share a variable scope? that would be a violation of the abstraction.

they can share an outer scope, though.

Reply:
I meant outer scope

Reply:
building in the same outer scope is good, but you still need each model to have a unique inner scope. there's no way around it without violating the abstraction.

if i already have a machine for bundling together all the variables inside a single scope (i.e. a `Model`), why would I start trying to look up those variables using scopes/names again? the whole point of making `Model` was to never do that again.

Reply:
can you send me a link to code showing how you solved this?